File: Leetcode/cp238_product_of_array_except_self/Solution.java
class Solution {
    // Time complexity O(n)
    // Space complexity O(n)
    public int[] productExceptSelf(int[] nums) {
        int[] res = new int[nums.length];
        res[0] = 1;
        // Calculate product from left
        for (int i = 1; i < nums.length; i++) {
            res[i] = res[i - 1] * nums[i - 1];
        }
        // Calculate right and product from the end of the array
        int right = 1;
        for (int i = nums.length - 1; i >= 0; i--) {
            res[i] *= right;
            right *= nums[i];
        }
        return res;
    }
}

File: Leetcode/cp238_product_of_array_except_self/Solution.py
import math

class Solution:
    # Leetcode Problem: https://leetcode.com/problems/product-of-array-except-self/
    #
    # Time complexity: O(n^2)
    # The first n is for the for loop and second n is form the math.prod function
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        result = []
        for idx in range(len(nums)):
            arr = nums[:idx] + nums[idx+1:]
            sum = math.prod(arr)
            result.append(sum)
        return result
